File: src/main/java/org/ffpy/plugin/coding/action/BaseAction.java
package org.ffpy.plugin.coding.action;

import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;
import com.intellij.openapi.project.DumbAware;
import lombok.extern.slf4j.Slf4j;
import org.ffpy.plugin.coding.action.env.ActionEnv;
import org.ffpy.plugin.coding.config.ActionEvnConfig;
import org.ffpy.plugin.coding.util.NotificationHelper;
import org.ffpy.plugin.coding.util.SpringContextUtils;
import org.jetbrains.annotations.NotNull;
import org.springframework.beans.factory.annotation.Autowired;

@Slf4j
public abstract class BaseAction extends AnAction implements DumbAware {

    public abstract void action() throws Exception;

    @Autowired
    protected ActionEnv env;

    @Override
    public void actionPerformed(@NotNull AnActionEvent event) {
        try {
            ActionEvnConfig.setEvent(event);
            SpringContextUtils.getBean(getClass()).action();
        } catch (Exception e) {
            log.error(e.getMessage(), e);
            NotificationHelper.error(e.getMessage()).show();
        } finally {
            ActionEvnConfig.clearEvent();
        }
    }
}

File: src/main/kotlin/org/ffpy/plugin/coding/action/BaseAction.kt
package org.ffpy.plugin.coding.action

import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent
import com.intellij.openapi.project.DumbAware
import org.ffpy.plugin.coding.util.NotificationHelper
import org.slf4j.Logger
import org.slf4j.LoggerFactory

abstract class BaseAction : AnAction(), DumbAware {

    companion object {
        val log: Logger = LoggerFactory.getLogger(BaseAction::class.java)
    }

    private val envThreadLocal = ThreadLocal<ActionEnv>()

    protected val env: ActionEnv
        get() = envThreadLocal.get()!!

    abstract fun action()

    override fun actionPerformed(e: AnActionEvent) {
        try {
            envThreadLocal.set(ActionEnv(e))
            action()
        } catch (e: Exception) {
            NotificationHelper.error(e.message).show()
            log.error(e.message, e)
        } finally {
            envThreadLocal.remove()
        }
    }
}